tests/test_tunnel.py

```diff
@@ -0,0 +1,151 @@
+
+    ecn_modes_map = {
+        "standard"       : "SAI_TUNNEL_DECAP_ECN_MODE_STANDARD",
+        "copy_from_outer": "SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER"
+    }
+
+    dscp_modes_map = {
+        "pipe"    : "SAI_TUNNEL_DSCP_MODE_PIPE_MODEL",
+        "uniform" : "SAI_TUNNEL_DSCP_MODE_UNIFORM_MODEL"
+    }
+
+    ttl_modes_map = {
+        "pipe"    : "SAI_TUNNEL_TTL_MODE_PIPE_MODEL",
+        "uniform" : "SAI_TUNNEL_TTL_MODE_UNIFORM_MODEL"
+    }
+
+
+    def check_interface_exists_in_asicdb(self, asicdb, sai_oid):
+        if_table = swsscommon.Table(asicdb, self.ASIC_RIF_TABLE)
+        status, fvs = if_table.get(sai_oid)
+        return status
+
+    def check_vr_exists_in_asicdb(self, asicdb, sai_oid):
+        vfr_table = swsscommon.Table(asicdb, self.ASIC_VRF_TABLE)
+        status, fvs = vfr_table.get(sai_oid)
+        return status
+
+    def check_tunnel_termination_entry_exists_in_asicdb(self, asicdb, tunnel_sai_oid, dst_ips):
+        tunnel_term_table = swsscommon.Table(asicdb, self.ASIC_TUNNEL_TERM_ENTRIES)
+
+        tunnel_term_entries = tunnel_term_table.getKeys()
+        assert len(tunnel_term_entries) == len(dst_ips)
+
+        for term_entry in tunnel_term_entries:
+            status, fvs = tunnel_term_table.get(term_entry)
+
+            for field, value in fvs:
+                if field == "SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_VR_ID":
+                    assert self.check_vr_exists_in_asicdb(asicdb, value)
+                elif field == "SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_TYPE":
+                    assert value == "SAI_TUNNEL_TERM_TABLE_ENTRY_TYPE_P2MP"
+                elif field == "SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_TUNNEL_TYPE":
+                    assert value == "SAI_TUNNEL_TYPE_IPINIP"
+                elif field == "SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_ACTION_TUNNEL_ID":
+                    assert value == tunnel_sai_oid
+                elif field == "SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_DST_IP":
+                    assert value in dst_ips
+                else:
+                    assert False, "Field %s is not tested" % field
+
+    def create_and_test_tunnel(self, db, asicdb, tunnel_name, src_ip,
+                               dst_ips, dscp_mode, ecn_mode, ttl_mode):
+        """ Create tunnel and verify all needed enties in ASIC DB exists """
+
+        # create tunnel entry in DB
+        ps = swsscommon.ProducerStateTable(db, self.APP_TUNNEL_DECAP_TABLE_NAME)
+        fvs = create_fvs(tunnel_type="IPINIP", src_ip=src_ip, dst_ip=",".join(dst_ips),
+                         dscp_mode=dscp_mode, ecn_mode=ecn_mode, ttl_mode=ttl_mode)
+
+        ps.set(tunnel_name, fvs)
+
+        # wait till config will be applied
+        time.sleep(1)
+
+        # check asic db table
+        tunnel_table = swsscommon.Table(asicdb, self.ASIC_TUNNEL_TABLE)
+
+        tunnels = tunnel_table.getKeys()
+        assert len(tunnels) == 1
+
+        tunnel_sai_obj = tunnels[0]
+
+        status, fvs = tunnel_table.get(tunnel_sai_obj)
```

Comment:
we probably need to assert status == true here

Reply:
Done